geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/geotools/GeoMesaDataStore.scala

```diff
@@ -586,6 +655,64 @@ object GeoMesaDataStore extends LazyLogging {
     }
   }
 
+  sealed trait SchemaCompatibility {
+
+    /**
+     * Ensures that the schema matches the existing schema in the data store, or throws an error if
+     * the schemas are incompatible
+     */
+    def apply(): Unit
+  }
+
+  object SchemaCompatibility {
+
+    /**
+     * Indicates that the schema is equal to the existing schema in the data store
+     */
+    case object Unchanged extends SchemaCompatibility {
+      override def apply(): Unit = {}
+    }
+
+    /**
+     * Indicates that the schema does not exist in the data store
+     *
+     * @param ds data store
+     * @param sft schema
+     */
+    class Missing(ds: GeoMesaDataStore[_], val sft: SimpleFeatureType) extends SchemaCompatibility {
+      override def apply(): Unit = ds.createSchema(sft)
+    }
+
+    object Missing {
+      def unapply(arg: Missing): Option[SimpleFeatureType] = Some(arg.sft)
+    }
+
+    /**
+     * Indicates that the schema is not equal to the existing schema, but is compatible
+     *
+     * @param ds data store
+     * @param typeName type name
+     * @param update the updated schema with all appropriate metadata
+     */
+    class Compatible(ds: GeoMesaDataStore[_], val typeName: String, val update: SimpleFeatureType)
+        extends SchemaCompatibility {
+      override def apply(): Unit = ds.updateSchema(typeName, update)
+    }
+
+    object Compatible {
+      def unapply(arg: Compatible): Option[(String, SimpleFeatureType)] = Some(arg.typeName, arg.update)
```

Comment:
Hehehe, I find the function name 'unapply' amusing here.  (I know it is to have a destructor or whatever, but still...:))